test/wasm2c/hello.txt

```diff
@@ -856,47 +857,46 @@ static void w2c__start(Z_test_instance_t* instance) {
   w2c_i2 = 1u;
   w2c_i3 = 0u;
   w2c_i4 = 0u;
-  w2c_i0 = CALL_INDIRECT(instance->w2c_T0, u32 (*)(void*, u32, u32, u32, u32), w2c_t0, w2c_i4, instance->w2c_T0.data[w2c_i4].module_instance, w2c_i0, w2c_i1, w2c_i2, w2c_i3);
-  (*Z_wasi_snapshot_preview1Z_proc_exit)(instance->Z_wasi_snapshot_preview1_instance, w2c_i0);
+  w2c_i0 = CALL_INDIRECT(instance->w2c_T0, u32 (*)(void*, u32, u32, u32, u32), w2c_test_t0, w2c_i4, instance->w2c_T0.data[w2c_i4].module_instance, w2c_i0, w2c_i1, w2c_i2, w2c_i3);
+  (*w2c_wasi0x5Fsnapshot0x5Fpreview1_proc_exit)(instance->w2c_wasi0x5Fsnapshot0x5Fpreview1_instance, w2c_i0);
   FUNC_EPILOGUE;
 }
 
 /* export: 'memory' */
-wasm_rt_memory_t* Z_testZ_memory(Z_test_instance_t* instance) {
+wasm_rt_memory_t* w2c_test_memory(w2c_test* instance) {
   return &instance->w2c_memory;
 }
 
 /* export: '_start' */
-void Z_testZ__start(Z_test_instance_t* instance) {
-  return w2c__start(instance);
+void w2c_test__start(w2c_test* instance) {
+  return w2c_test__start_0(instance);
 }
 
-static void init_instance_import(Z_test_instance_t* instance, struct Z_wasi_snapshot_preview1_instance_t* Z_wasi_snapshot_preview1_instance){
-  instance->Z_wasi_snapshot_preview1_instance = Z_wasi_snapshot_preview1_instance;
+static void init_instance_import(w2c_test* instance, struct w2c_wasi0x5Fsnapshot0x5Fpreview1* w2c_wasi0x5Fsnapshot0x5Fpreview1_instance){
+  instance->w2c_wasi0x5Fsnapshot0x5Fpreview1_instance = w2c_wasi0x5Fsnapshot0x5Fpreview1_instance;
 }
-
-void Z_test_instantiate(Z_test_instance_t* instance, struct Z_wasi_snapshot_preview1_instance_t* Z_wasi_snapshot_preview1_instance) {
+void wasm2c_test_instantiate(w2c_test* instance, struct w2c_wasi0x5Fsnapshot0x5Fpreview1* w2c_wasi0x5Fsnapshot0x5Fpreview1_instance) {
   assert(wasm_rt_is_initialized());
-  init_instance_import(instance, Z_wasi_snapshot_preview1_instance);
+  init_instance_import(instance, w2c_wasi0x5Fsnapshot0x5Fpreview1_instance);
```

Comment:
Given how common underscores are likely to be in module names this seems like a somewhat of a regression.

I'm not sure what we can do about?  Some other kind of escaping mechanism for modules with underscores in their name?  Use double underscores to mark the start and end of the module name?

Reply:
We can definitely prettify this at the expense of making the mangling rules more complicated.

One possibility would be: In all names, we escape leading underscores, trailing underscores, and underscores that follow underscores. Additionally, in a module name, we transform a single underscore into a double underscore. The boundary between module name and export name is the first single underscore after `w2c_`.

So `wasi_snapshot_preview1` would become `wasi__snapshot__preview1`. Unfortunately an export named `_start` would become `0x5Fstart` which is a little uglier but maybe still acceptable?

Reply:
Just pushed a new commit that does this -- hope it looks cleaner.